assets/src/edit-story/components/library/panes/media/mediaPane.js

```diff
@@ -226,17 +227,76 @@ function MediaPane(props) {
     return null;
   };
 
+  /**
+   * Insert element such image, video and audio into the editor.
+   *
+   * @param {Object} attachment Attachment object
+   * @param {number} width      Width that element is inserted into editor.
+   * @return {null|*}          Return onInsert or null.
+   */
+  const createMediaElement = (attachment, width) => {
+    const { src, mimeType, oWidth, oHeight } = attachment;
+    const origRatio = oWidth / oHeight;
+    const height = width / origRatio;
+    if (allowedImageMimeTypes.includes(mimeType)) {
+      return createElementDef('image', {
+        src,
+        width,
+        height,
+        x: 5,
+        y: 5,
+        rotationAngle: 0,
+        origRatio,
+        origWidth: oWidth,
+        origHeight: oHeight,
+      });
+    } else if (allowedVideoMimeTypes.includes(mimeType)) {
+      const { id: videoId, poster, posterId: posterIdRaw } = attachment;
+      const posterId = parseInt(posterIdRaw);
+      const videoEl = createElementDef('video', {
+        src,
+        width,
+        height,
+        x: 5,
+        y: 5,
+        rotationAngle: 0,
+        origRatio,
+        origWidth: oWidth,
+        origHeight: oHeight,
+        mimeType,
+        videoId,
+        posterId,
+        poster,
+      });
+
+      // Generate video poster if one not set.
+      if (videoId && !posterId) {
+        uploadVideoFrame(videoId, src, videoEl.id);
+      }
+
+      return videoEl;
+    }
+    return null;
+  };
+
   /**
    * Get a formatted element for different media types.
    *
    * @param {Object} mediaEl Attachment object
    * @param {number} width      Width that element is inserted into editor.
    * @return {null|*}          Element or null if does not map to video/image.
    */
-  const getMediaElement = (mediaEl, width) => {
+  const MediaElement = ({ mediaEl, width }) => {
+    const element = createMediaElement(mediaEl, width);
```

Comment:
Seems like it's a bit early to create an actual element? There are all kind of things that we will be adding for element creation eventually. So it could be getting a bit heavier and heavier. Some other nuances:
 * Would it be reasonable to add the same image twice? I think there was even one use case with image splitting that would rely on that.
 * The `uploadVideoFrame` above should probably run in a side-effect or an event handler. Not on rendering.

Reply:
I cleaned up the implementation a bit but I still think that this abstraction is very useful so that we can have the same code for merging elements both coming from the gallery and coming from the canvas. Let's discuss more if needed.